src/Eris/Generator/GeneratedValue.php

```diff
@@ -0,0 +1,50 @@
+<?php
+namespace Eris\Generator;
+
+use Eris\Generator\GeneratedValue;
+use InvalidArgumentException;
+
+final class GeneratedValue
+{
+    private $value;
+    private $input;
+    private $generatorName;
+    
+    public static function fromValueAndInput($value, $input, $generatorName = null)
+    {
+        if ($value instanceof self) {
+            throw new InvalidArgumentException("It looks like you are trying to build a GeneratedValue whose value is another GeneratedValue. This is almost always an error as values will be passed as-is to properties and GeneratedValue should be hidden from them.");
+        }
+        return new self($value, $input, $generatorName);
+    }
+
+    public static function fromJustValue($value, $generatorName = null)
+    {
+        return new self($value, $value, $generatorName);
+    }
+    
+    private function __construct($value, $input, $generatorName)
+    {
+        if ($value instanceof self) {
+            throw new InvalidArgumentException("It looks like you are trying to build a GeneratedValue whose value is another GeneratedValue. This is almost always an error as values will be passed as-is to properties and GeneratedValue should be hidden from them.");
+        }
```

Comment:
This is a check that $value is `not` of that type, to prevent accidental nesting